metashare/api/jobs.py

```diff
@@ -0,0 +1,191 @@
+import contextlib
+import logging
+import traceback
+
+from asgiref.sync import async_to_sync
+from cumulusci.core import flowrunner
+from cumulusci.core.config import ScratchOrgConfig
+from django.utils.text import slugify
+from django.utils.timezone import now
+from django_rq import job
+from github3 import login
+from github3.exceptions import UnprocessableEntity
+
+from .custom_cci_configs import MetaShareCCI
+from .github_context import (
+    extract_owner_and_repo,
+    get_cumulus_prefix,
+    local_github_checkout,
+)
+from .push import push_message_about_instance
+
+logger = logging.getLogger(__name__)
+
+
+async def report_scratch_org_error(instance, message):
+    from .serializers import ScratchOrgSerializer
+
+    message = {
+        "type": "SCRATCH_ORG_PROVISION_FAILED",
+        "payload": {"message": message, "model": ScratchOrgSerializer(instance).data},
+    }
+    await push_message_about_instance(instance, message)
+
+
+@contextlib.contextmanager
+def report_errors_on(scratch_org):
+    try:
+        yield
+    except Exception as e:
+        async_to_sync(report_scratch_org_error)(scratch_org, str(e))
+        tb = traceback.format_exc()
+        logger.error(tb)
+        scratch_org.delete()
+        raise
+
+
+def create_scratch_org(*, scratch_org, user, repo_url, commit_ish):
+    """
+    Expects to be called in the context of a local github checkout.
+    """
+    gh = login(token=user.gh_token)
+    owner, repo = extract_owner_and_repo(repo_url)
+    repository = gh.repository(owner, repo)
+    branch = repository.branch(commit_ish)
+    commit = branch.commit
+    latest_commit = commit.sha
+    latest_commit_url = commit.html_url
+    latest_commit_at = commit.commit.author.get("date", None)
+
+    with report_errors_on(scratch_org):
+        org_config = ScratchOrgConfig(
+            {"config_file": "orgs/dev.json", "scratch": True}, "dev"
+        )
+        org_config.create_org()
+        # We just need to access this property to pull down info from SF
+        # as a side-effect.
+        org_config.scratch_info
+        # TODO: We should actually just save all of org_config.config as
+        # json to the model, and use it to re-hydrate the
+        # ScratchOrgConfig as-needed.
+        scratch_org.url = org_config.instance_url
+        scratch_org.last_modified_at = now()
+        scratch_org.expires_at = org_config.expires
+        scratch_org.latest_commit = latest_commit
+        scratch_org.latest_commit_url = latest_commit_url
+        scratch_org.latest_commit_at = latest_commit_at
+        scratch_org.config = org_config.config
+        scratch_org.save()
```

Comment:
This triggers the success WS event, but I think we want to wait until the `dev_org` flow is done too.

Reply:
Thanks, fixed.